towerlib/towerlib.py

```diff
@@ -1997,24 +2011,579 @@ def schedules(self):
+        return required_job_templates
+
+    def change_job_template_data(self, job_template_data):
+        """Send API PATCH request to update the job template information with the given data.
+        https://docs.ansible.com/ansible-tower/3.6.1/html/towerapi/api_ref.html#/Authentication/Authentication_applications_partial_update_0
+
+        Args:
+            job_template_data: updated data of the job template as dictionary
+
+        Returns:
+            list: List of response of api request as json on success, False otherwise.
+
+        """
+        job_template_id = job_template_data['id']
+        job_url = '{api}/job_templates/{id}/'.format(api=self.api, id=job_template_id)
+        response = self.session.patch(job_url, data=json.dumps(job_template_data))
+        if not response.ok:
+            self._logger.error("Error updating the job template with the given data '{}'".format(job_template_data))
+            return None
+        else:
+            self._logger.info("Job template with the name '{}' is updated successfully.".format(job_template_data['name']))
+            return response.json()
+
+    def change_project_of_job_template(self, job_template, project_id):
+        """A job template in the ansible tower has to have a project from the list of projects.
+
+        This function helps to change the project name of a job template in ansible tower for a given job template.
+
+        Args:
+            job_template: given job template object
+            project_id: given project's id
+
+        Returns:
+            Object: The updated job_template object
+
+        """
+
+        project_obj = self.get_project_by_id(project_id)
+        project_playbooks = self.get_playbooks_by_branch_name(project_obj.scm_url, project_obj.scm_branch)
+        new_playbook = job_template.playbook
+        if new_playbook not in project_playbooks:
+            self._logger.warning("Job template's playbook '{}' is not found in the given project's playbook"
+                                 "list. Now changing the playbook to '{}'. Please select a playbook manually."
+                                 .format(new_playbook, project_playbooks[0]))
+            new_playbook = project_playbooks[0]
+
+        new_data = {
+            "id": job_template.id,
+            "project": project_obj.id,
+            "playbook": new_playbook,
+            "name": job_template.name,
+            "description": job_template.description
+        }
+        response = self.change_job_template_data(new_data)
+        if response is not None:
+            self._logger.info("The project for the job template '{}' is changed from '{}' to '{}'".
+                              format(job_template.name, job_template.project.name, project_obj.name))
+        else:
+            self._logger.error("Error updating the job template's project.")
+        return response
+
+    def change_project_of_job_template_to_prod_branch_project(self, job_template, scm_url):
```

Comment:
I think that this is client side business logic and should not be part of the library. The library should expose the capability to update the project name of a job template but nothing more opinionated than that. The job template should have setters for the attributes that need to be updated.

Reply:
Removed this method.